src/psyclone/tests/psyad/domain/lfric/test_lfric_adjoint_harness.py

```diff
@@ -619,6 +619,8 @@ def test_generate_lfric_adjoint_harness(fortran_reader, fortran_writer):
             "    inner2 = inner2 + field_field_input_inner_prod\n" in gen)
 
 
+@pytest.mark.xfail(reason="func_type and gh_quadrature_xyoz are neither "
```

Comment:
Why has this test started to fail now? It seems as though it's unrelated to this PR?

Reply:
There are both `extends` and `procedure` in `TL_CODE`, so this was defaulting to a `DataTypeSymbol` of `UnsupportedFortranType` before this PR.

The error if not xfailing this comes from the call to `generate_lfric_adjoint_harness` and is:
`psyclone.psyir.backend.visitor.VisitorError: Visitor Error: The following symbols are not explicitly declared or imported from a module and there are no wildcard imports which could be bringing them into scope: 'func_type', 'gh_quadrature_xyoz'`

`new_code` (after replacement) is:
```fortran
module testkern_mod
  use kinds_mod, only: i_def, r_def
  use kernel_mod, only: kernel_type, arg_type, gh_field, gh_real, gh_write, w3, cell_column
  type, extends(kernel_type) :: testkern_type
     type(arg_type), dimension(2) :: meta_args =   & 
          (/ arg_type(gh_scalar, gh_real, gh_read),       & 
      arg_type(gh_field,  gh_real, gh_write,  w3)  & 
    /)
     type(func_type) :: meta_funcs(1) = (/        &
          func_type(W3, gh_basis)          &
          /)
     integer :: operates_on = cell_column
     integer :: gh_shape = gh_quadrature_xyoz
   contains
     procedure, nopass :: code => testkern_code
  end type testkern_type
       contains
  subroutine testkern_code(nlayers, ascalar, field, ndf_w3, undf_w3, map_w3, basis_w3_qr_xyoz, np_xy_qr_xyoz, np_z_qr_xyoz, weights_xy_qr_xyoz, weights_z_qr_xyoz)
      INTEGER(KIND=i_def), intent(in) :: np_xy_qr_xyoz, np_z_qr_xyoz
      REAL(KIND=r_def), intent(in), dimension(1,ndf_w3,np_xy_qr_xyoz,np_z_qr_xyoz) :: basis_w3_qr_xyoz
      REAL(KIND=r_def), intent(in), dimension(np_xy_qr_xyoz) :: weights_xy_qr_xyoz
      REAL(KIND=r_def), intent(in), dimension(np_z_qr_xyoz) :: weights_z_qr_xyoz
    integer(kind=i_def), intent(in) :: nlayers
    integer(kind=i_def), intent(in) :: ndf_w3, undf_w3
    integer(kind=i_def), intent(in), dimension(ndf_w3) :: map_w3
    real(kind=r_def), intent(in) :: ascalar
    real(kind=r_def), intent(inout), dimension(undf_w3) :: field
    field = ascalar
  end subroutine testkern_code
end module testkern_mod
```
and indeed, neither `func_type` nor `gh_quadrature_xyoz` are imported.

Should I add them to `TL_CODE` or is the string replace missing something?

Reply:
Aha! It's obvious now you've explained it :-) Yes, please add `use argument_mod, only: func_type, gh_quadrature_xyoz` to TL_CODE .

Reply:
All good now!